crates/trie/src/proof.rs

```diff
@@ -55,226 +40,196 @@ where
 
-impl<'a, 'tx, TX, H> ProofRestorer<'a, TX, H>
-where
-    TX: DbTx<'tx>,
-    H: HashedCursorFactory + Clone,
-{
-    /// Set the hashed cursor factory.
-    fn with_hashed_cursor_factory<HF>(
-        self,
-        hashed_cursor_factory: HF,
-    ) -> Result<ProofRestorer<'a, TX, HF>, ProofError>
-    where
-        HF: HashedCursorFactory,
-    {
-        let hashed_account_cursor = hashed_cursor_factory.hashed_account_cursor()?;
-        Ok(ProofRestorer {
-            tx: self.tx,
-            hashed_cursor_factory,
-            hashed_account_cursor,
-            account_rlp_buf: self.account_rlp_buf,
-            node_rlp_buf: self.node_rlp_buf,
-        })
-    }
+        let mut hash_builder = HashBuilder::default().with_proof_retainer(target_nibbles);
+        while let Some(key) = walker.key() {
+            if walker.can_skip_current_node {
+                hash_builder.add_branch(key, walker.hash().unwrap(), walker.children_are_in_trie());
+            }
 
-    fn restore_branch_node(
-        &mut self,
-        prefix: &Nibbles,
-        node: BranchNodeCompact,
-    ) -> Result<Bytes, ProofError> {
-        let mut hash_idx = 0;
-        let mut branch_node_stack = Vec::with_capacity(node.state_mask.count_ones() as usize);
-
-        for child in CHILD_INDEX_RANGE.filter(|ch| node.state_mask.is_bit_set(*ch)) {
-            if node.hash_mask.is_bit_set(child) {
-                branch_node_stack.push(rlp_hash(node.hashes[hash_idx]));
-                hash_idx += 1;
-            } else {
-                let child_key = prefix.join(&Nibbles::from_hex(Vec::from([child])));
-                let mut child_key_to_seek = child_key.pack();
-                child_key_to_seek.resize(32, 0);
-
-                let leaf_node_rlp =
-                    self.restore_leaf_node(B256::from_slice(&child_key_to_seek), child_key.len())?;
-                branch_node_stack.push(leaf_node_rlp.to_vec());
+            let seek_key = match walker.next_unprocessed_key() {
+                Some(key) => key,
+                None => break, // no more keys
+            };
+
+            let next_key = walker.advance()?;
+            let mut storage = hashed_storage_cursor.seek(hashed_address, seek_key)?;
+            while let Some(StorageEntry { key: hashed_key, value }) = storage {
+                let hashed_key_nibbles = Nibbles::unpack(hashed_key);
+                if let Some(ref key) = next_key {
+                    if key < &hashed_key_nibbles {
+                        break
+                    }
+                }
+
+                if let Some(proof) =
+                    proofs.iter_mut().find(|proof| proof.nibbles == hashed_key_nibbles)
+                {
+                    proof.set_value(value);
+                }
+
+                hash_builder
+                    .add_leaf(hashed_key_nibbles, alloy_rlp::encode_fixed_size(&value).as_ref());
+                storage = hashed_storage_cursor.next()?;
```

Comment:
more of a general question, is there no good way to avoid some of the duplication between this code and `calculate` methods?

Reply:
have been thinking about this as well, perhaps this could be an iterator

Reply:
the refactor will touch the core trie calc, so it's better to do it in a separate PR